plugin/pkg/scheduler/algorithm/priorities/service_affinity.go

```diff
@@ -0,0 +1,98 @@
+/*
+Copyright 2014 The Kubernetes Authors All rights reserved.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package priorities
+
+import (
+	"github.com/golang/glog"
+	"k8s.io/kubernetes/pkg/api"
+	"k8s.io/kubernetes/pkg/labels"
+	"k8s.io/kubernetes/plugin/pkg/scheduler/algorithm"
+)
+
+type ServiceAffinity struct {
+	serviceLister algorithm.ServiceLister
+}
+
+func NewServiceAffinityPriority(serviceLister algorithm.ServiceLister) algorithm.PriorityFunction {
+	serviceAffinity := &ServiceAffinity{
+		serviceLister: serviceLister,
+	}
+	return serviceAffinity.CalculateAffinityPriority
+}
+
+// CalculateAffinityPriority affiliates a pod to the existing services' pods
+// that matches the labels indicated in affinitySelector.
+// The more existing pods to affiliate deployed on, the higher priority the node gets.
+func (s *ServiceAffinity) CalculateAffinityPriority(pod *api.Pod, podLister algorithm.PodLister, nodeLister algorithm.NodeLister) (algorithm.HostPriorityList, error) {
+	var maxCount int
+	counts := map[string]int{}
+	affinitySelector := labels.Set(pod.Spec.AffinitySelector)
+
+	// Logically, we should match pod's affinitySelector with services' selectors.
+	// Then, to find services' pods, we need to match pods' labels with services'
+	// selectors. Finally, calculate the priorties with affinitySelector and the
+	// found pods' labels.
+	// As a result of optimization, we match the pod's affinitySelector with
+	// existing pods' labels here.
+	allPods, err := podLister.List(labels.Everything())
+	if err != nil {
+		glog.V(10).Infof("PodLister Error")
+		return nil, err
+	}
+
+	if len(allPods) > 0 {
+		for _, onePod := range allPods {
+			// Only match pods in the same namespace
+			if onePod.Namespace != pod.Namespace {
+				continue
+			}
+
+			// match affinitySelector with every pod's labels.
+			// Every matched label will add to the node's priority
+			for key, val := range onePod.ObjectMeta.Labels {
```

Comment:
Right, we did a bit hacky implement here to support affiliate pod to multiple services.
Actually, the affinity relation to multiple services should be indicated in a selector array.

A more general way, if also want to support nodeSelector , maybe defining a data structure like this:
```go
type affinityEntity struct{
    // indicate the type of target to affiliate, RC, Pod, Service, Node e.g.
    TargetType string, `json:"targetType"`
    // indicate hard or soft affinity
    AffinityType string, `json:"affinityType"`
    Selector map[string]string, `json:"selector"`
}

type PodSpec struct {
    ...
    affinitySelector []affinityEntity,
    ...
}
```
then, an example for user to use would like:
```yaml
apiVersion: v1
kind: Pod
metadata:
  labels:
    name: nodejs
spec:
  containers:
  - name: nodejs-container
    image: nodejs
  affinitySelector:
  - targetType: Service
    affinityType: soft
    selector:
      name: redis
  - targetType: Node
    affinityType: hard
    selector:
      volumeType: SSD
```